packages/runtime/src/next/api-handler/handlers/proxy-preview-mode.ts

```diff
@@ -85,71 +112,53 @@ async function proxyPreviewModeApiRouteHandler(
-  }
-
-  const isForwardedProtoHttps = forwardedProto.includes('https')
-
-  const forwardedSSL = req.headers['x-forwarded-ssl']
-  const isForwardedSSL = typeof forwardedSSL === 'string' && forwardedSSL === 'on'
-
-  const proto = isForwardedProtoHttps || isForwardedSSL ? 'https' : 'http'
-  let target = `${proto}://${host}`
-
-  // If the user generates a locally-trusted CA for their SSL cert, it's likely that Node.js won't
-  // trust this CA unless they used the `NODE_EXTRA_CA_CERTS` option
-  // (see https://stackoverflow.com/a/68135600). To provide a better developer experience, instead
-  // of requiring the user to provide the CA to Node.js, we don't enforce SSL during local
-  // development.
-  const secure = process.env['NODE_ENV'] === 'production'
+  proxyUrl.searchParams.delete('x-makeswift-preview-mode')
+  proxyHeaders.delete('x-makeswift-preview-mode')
+  // The following headers are Next.js-specific and are removed to prevent Next.js from
+  // short-circuiting requests, breaking routing.
+  proxyHeaders.delete('x-invoke-path')
+  proxyHeaders.delete('x-invoke-query')
 
   const previewData: MakeswiftPreviewData = {
     makeswift: true,
+    // This will eventually be dynamic
     siteVersion: MakeswiftSiteVersion.Working,
   }
-  const setCookie = res.setPreviewData(previewData).getHeader('Set-Cookie')
-  res.removeHeader('Set-Cookie')
+
+  const setCookie = res.setPreviewData(previewData).getHeader('set-cookie')
+  res.removeHeader('set-cookie')
 
   if (!Array.isArray(setCookie)) return res.status(500).send('Internal Server Error')
 
-  const additionalCookies = parse(setCookie)
-    .map(cookie => serialize(cookie.name, cookie.value, cookie as CookieSerializeOptions))
-    .join(';')
+  setCookie.forEach((cookie) => proxyHeaders.append('cookie', cookie))
+
+  const response = await fetch(proxyUrl, {
+    referrer: req.headers.referer,
+    headers: proxyHeaders,
+  });
+
+  response.headers.forEach((value, name) => {
+    res.setHeader(name, value);
+  })
+
+  res.statusCode = response.status;
+  res.statusMessage = response.statusText
+
+  // `fetch` automatically decompresses the response, but the response headers will keep the
+  // `content-encoding` and `content-length` headers. This will cause decoding issues if the client
+  // attempts to decompress the response again. To prevent this, we remove these headers.
+  //
+  // See https://github.com/nodejs/undici/issues/2514.
+  if (res.hasHeader('content-encoding')) {
+    res.removeHeader('content-encoding')
+    res.removeHeader('content-length')
+  }
 
-  const cookie = originalCookies ? `${originalCookies}; ${additionalCookies}` : additionalCookies
+  const arrayBuffer = await response.arrayBuffer()
 
-  return await new Promise<void>((resolve, reject) =>
-    previewModeProxy.web(req, res, { target, headers: { cookie }, secure }, err => {
-      if (err) reject(err)
-      else resolve()
-    }),
-  )
+  res.write(arrayBuffer)
+  res.end()
```

Comment:
I think we can consolidate these two lines to `res.send(arrayBuffer)`

Reply:
It doesn't actually work. It treats the `arrayBuffer` as an `application/json` mime type.

Reply:
Oh, interesting. Well, what you have is perfectly good.